src/Formats/SchemaInferenceUtils.cpp

```diff
@@ -649,6 +666,11 @@ namespace
 
             /// Check settings specific for JSON formats.
 
+            if (settings.json.infer_variant_from_multi_type_array)
+            {
+                transformVariant(data_types, type_indexes);
+            }
```

Comment:
This check sholuld be the last one in this method, not the first one. We could change the data types in other methods below. And we should do it not only for JSON but for all types